feat(api): Eager chat template warmup to eliminate first-request latency

TheCodeWrangler opened the pull request:
## Summary
Adds automatic warmup of chat template processing during server initialization to eliminate the significant latency penalty on the first chat completion request.

## Problem
When a vLLM server starts, the first chat completion request experiences much higher latency than subsequent requests. This is caused by lazy initialization of:

1. **Chat template content format auto-detection** — Analyzes the Jinja2 template structure
2. **Jinja2 template compilation** — Compiles the chat template on first use  
3. **Tokenizer chat initialization** — First-time setup of tokenizer for chat formatting

This creates problems in production:
- Load balancers may timeout on the first request
- Users experience inconsistent response times
- The `/health` endpoint reports healthy before the server is truly "warm"

## Solution
Add a `warmup()` method to `OpenAIServingChat` that runs during `init_app_state()`, **before** the server starts accepting requests:

```python
async def warmup(self) -> None:
    # Get tokenizer and create minimal dummy request
    tokenizer = await self.engine_client.get_tokenizer()
    dummy_request = ChatCompletionRequest(
        messages=[{"role": "user", "content": "warmup"}],
        model=None,
        max_completion_tokens=1,
    )
    # Trigger template compilation
    await self._preprocess_chat(dummy_request, tokenizer, ...)
```

**Key design decisions:**
- Reuses existing `_preprocess_chat()` rather than duplicating logic
- Handles exceptions gracefully (logs warning with full traceback, doesn't fail startup)
- Logs timing for observability
- Runs before `/health` reports healthy

## Changes
- `vllm/entrypoints/openai/serving_chat.py` — Added `warmup()` method (+49 lines)
- `vllm/entrypoints/openai/api_server.py` — Call warmup in `init_app_state()` (+3 lines)

---

## Verification

Tested on **Qwen/Qwen3-4B** on **NVIDIA H100** with **vllm/vllm-openai:v0.12.0**.

*Note: Actual timings vary based on model, hardware, and configuration.*

**Before (without warmup):**
| Request | Latency |
|---------|---------|
| 1st chat completion | ~1050ms |
| 2nd chat completion | ~170ms |

**After (with warmup):**
| Request | Latency |
|---------|---------|
| 1st chat completion | ~170ms |
| 2nd chat completion | ~170ms |

Server logs now show warmup during initialization:
```
INFO: Warming up chat template processing...
INFO: Detected the chat template content format to be 'string'...
INFO: Chat template warmup completed in ~850ms
```

The warmup cost is moved from the first user request to server startup, ensuring consistent response times from the first request onward.

TheCodeWrangler commented:
@LucasWilkinson @robertgshaw2-redhat 

Updated implementation prior (now closed) PR regarding first call latency
https://github.com/vllm-project/vllm/pull/30561

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -162,6 +162,55 @@ def __init__(
+
+        except Exception:
+            # Log but don't fail server startup if warmup fails
+            logger.warning("Chat template warmup failed:", exc_info=True)
```
I would increase the log level and use `logger.exception`

TheCodeWrangler committed:
feat(api): eager chat template warmup to eliminate first-request latency

- Add warmup() method to OpenAIServingChat to pre-compile Jinja2 templates
- Call warmup() during app initialization to avoid first-request latency
- Warmup triggers template compilation and content format detection
- Reduces first request latency by pre-initializing chat template processing

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Remove kernel warmup changes

Signed-off-by: Nathan Price <nathan@abridge.com>

DarkLight1337 reviewed (approved):
Thanks, LGTM now

TheCodeWrangler committed:
Merge branch 'main' into feature/eager-chat-template-init

TheCodeWrangler committed:
Merge branch 'main' into feature/eager-chat-template-init
